server.py

```diff
@@ -79,17 +142,26 @@ def home():
 @app.get("/dashboard")
 @login_required
 def dashboard():
+    contests = Contest.select().order_by(Contest.start_time)
     return bottle.template("dashboard.html", contests=contests)
 
 
 @app.get("/contest/<code>/<number>")
 @login_required
-def contest(code, number):
-    if not code in contests:
-        return "Contest does not exist"
-    if contests[code].start_time > datetime.datetime.now():
+def question(code, number):
+    if (
+        not ContestProblems.select()
+        .where((Contest.code == code) & (Question.q_no == int(number)))
+        .join(Contest, on=(ContestProblems.contest == Contest.id))
```

Comment:
these two can be skipped. If the contest exists it exists :smile:

Reply:
Let's have this one for now, just to make it more secure :)

Reply:
:thinking: how does it make it more secure? Lemme see:
- ContestProblems.contest == Contest.id
- ContestProblems.question == Question.id
- we could just check if the contest problem exists.

Ah!. you're doing a custom `q_no` as id business. Yep. You need a join for that. The contest join can exist/not-exist since it's not adding anything to the query.

Let's just keep it for now, we'll correct this when cleaning up the code